zulip_bots/zulip_bots/bots/trivia_quiz/trivia_quiz.py

```diff
@@ -198,16 +200,41 @@ def format_quiz_for_markdown(quiz_id, quiz):
     )
     return content
 
-def grade_question(quiz, answer):
-    correct = (answer == quiz['correct_letter'])
+def update_quiz(quiz: Dict[str, Any], quiz_id: str, bot_handler: Any) -> None:
+    bot_handler.storage.put(quiz_id, json.dumps(quiz))
 
-    if correct:
-        long_answer = quiz['answers'][answer]
-        response = '**CORRECT!** {long_answer} :tada:'.format(long_answer=long_answer)
-        return correct, response
+def handle_answer(quiz: Dict[str, Any], answer: str, quiz_id: str,
+                  bot_handler: Any) -> Tuple[bool, str]:
+    long_answer = quiz['answers'][quiz['correct_letter']]
+    start_new_question = False
+
+    if answer not in quiz['answered_options'] and not quiz['ended']:
+        # The question has not been ended, so we store all the options
+        # which are being answered
+        quiz['answered_options'].append(answer)
+        update_quiz(quiz, quiz_id, bot_handler)
+
+    if answer == quiz['correct_letter']:
+        response = '**CORRECT!** {long_answer} :tada:'
+        # We don't start a new question if the question has already been
+        # answered
+        start_new_question = not quiz['ended']
+    else:
+        if len(quiz['answered_options']) == 3:
+            # We reveal the answer and start a new question upon having three failed
+            # attempts at answering a question
+            response = '''Three failed attempts at {quiz_id}:disappointed:
+The correct answer is {long_answer}.'''
+            start_new_question = True
+            # we reset this so that the users can keep answering a question after it has ended
+            quiz['answered_options'] = []
+            quiz['ended'] = True
+            update_quiz(quiz, quiz_id, bot_handler)
+        else:
+            response = '**WRONG!** {answer} is not correct :disappointed:'
 
-    response = '**WRONG!** {answer} is not correct :disappointed:'.format(answer=answer)
-    return correct, response
+    response = response.format(answer=answer, long_answer=long_answer, quiz_id=quiz_id)
```

Comment:
MINOR: I appreciate this is long and maybe it wouldn't fit on one line, but I would combine these two lines either:
* splitting the two tuple components, if the line is overly long
* shorten the format keys or values slightly